### PR TITLE
Add TSNE Summary Module

### LINKED ISSUE #229: TSNE Summary
**What is your feature request related to?**
* Module request

**Please describe the module/problem**

I would like to have a TSNE SummaryModule. This TSNE SummaryModule should take multiple columns as input and provide corresponding tsne visualisations.

**Describe the solution you'd like**
A TSNESummary Module

**Describe alternatives you've considered**
With skTimeWrapper this is not possible, because it does not plot the resulting data. And masking inputs would also be not possible.

### PR DESCRIPTION
**Pull request by benHeid**
## Related Issues
closes #229 
## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos)
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] Did you update the CHANGELOG

## PR review
Anyone in the community is free to review the PR once the tests have passed.

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones to the PR so it can be classified

### CODE REVIEW

File: pywatts/summaries/tsne_visualisation.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+from pywatts.core.base_summary import BaseSummary
+from sklearn.manifold import TSNE
+import numpy as np
+
+from pywatts.core.summary_object import SummaryObjectList
+
+
+class TSNESummary(BaseSummary):
+    """
+    Summary that performs a TSNE to visualise the data. It is possible to specify mask by providing kwargs
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:15]: I would once write it out and would use t-SNE instead of TSNE:
Summary that performs a t-distributed stochastic neighbor embedding (t-SNE) to visualise the data. It is possible to specify masks by providing kwargs

File: pywatts/summaries/tsne_visualisation.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    to the transform method that end with _masked.
+    :param max_point: The maximum number of points per data that should be plotted
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:17]: data might be confusing.
data set?

File: pywatts/summaries/tsne_visualisation.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    :type max_point: int
+    :param all_in_one_plot: Flag indicating if all input data should be visualised in the same plot. If not the column
+                            GT is visualised with all other in separate plots.
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:20]: If not the column
GT is visualised like all other in separate plots.

File: pywatts/summaries/tsne_visualisation.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    :type all_in_one_plot: Boolean
+    :param tsne_params: Params for the TSNE Visualisation. See sklearn.
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:22]: Params for the t-SNE visualisation (see sklearn at https://scikit-learn.org/stable/modules/generated/sklearn.manifold.TSNE.html).

File: pywatts/summaries/tsne_visualisation.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                 max_points=10000, all_in_one_plot=False, tsne_params=None):
+        super().__init__(name)
+        self.tsne_params = {} if tsne_params is None else tsne_params
+        self.max_points = max_points
+        self.all_in_one_plot = all_in_one_plot
+
+    def get_params(self) -> Dict[str, object]:
+        """"
+        Get the params of the TSNE Summary
+        :return: A dict containing all parameter of the tsne_summary
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:36]: A dict containing all parameters of the TSNE Summary

File: pywatts/summaries/tsne_visualisation.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        """
+        return {
+            "tsne_params": self.tsne_params,
+            "max_points": self.max_points,
+            "all_in_one_plot": self.all_in_one_plot
+        }
+
+    def set_params(self, max_points=None, all_in_one_plot=None, tsne_params=None):
+        """
+        Set the params of the tsne summary.
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:47]: Set the params of the TSNE Summary

File: pywatts/summaries/tsne_visualisation.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        :param max_point: The maximum number of points per data that should be plotted
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:48]: see above

File: pywatts/summaries/tsne_visualisation.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        :type max_point: int
+        :param all_in_one_plot: Flag indicating if all input data should be visualised in the same plot. If not the column
+                                GT is visualised with all other in separate plots.
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:51]: see above

File: pywatts/summaries/tsne_visualisation.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        :type all_in_one_plot: Boolean
+        :param tsne_params: Params for the TSNE Visualisation. See sklearn.
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:53]: see above

File: pywatts/summaries/tsne_visualisation.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+            self.max_points = max_points
+        if all_in_one_plot is not None:
+            self.all_in_one_plot = all_in_one_plot
+        if tsne_params is not None:
+            self.tsne_params = tsne_params
+
+    def transform(self, file_manager, gt: xr.DataArray, **kwargs: xr.DataArray):
+        """
+        Calculates the TSNE and visualise it. kwargs that end with _masked are masks for the other input. E.g. gt_masked
+        is a mask for gt.
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:66]: Calculates the TSNE and visualises it. kwargs that end with _masked are masks for the other input. For example, gt_masked is a mask for gt.

File: pywatts/summaries/tsne_visualisation.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        tsne = TSNE(**self.tsne_params)
+        data_final = [gt_data]
+        data_final.extend(kwargs_data.values())
+        result = tsne.fit_transform(np.concatenate(data_final))
+        plt.scatter(result[:len(gt_data), 0], result[:len(gt_data), 1], alpha=0.5, label="Ground Truth")
+        for i, key in enumerate(kwargs_data, 1):
+            plt.scatter(result[len(gt_data) * i:len(gt_data) * (i + 1), 0],
+                        result[len(gt_data) * i:len(gt_data) * (i + 1), 1], alpha=0.5,
+                        label=key)
+        plt.xlabel("x-tsne")
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:101]: plt.xlabel("x-t-SNE")

File: pywatts/summaries/tsne_visualisation.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        plt.ylabel("y-tsne")
```
Comment (marianturowski) [inline pywatts/summaries/tsne_visualisation.py:102]: plt.ylabel("y-t-SNE")

**Review by marianturowski — CHANGES_REQUESTED** [review]
The implementation of the t-SNE Summary looks good. I just recommend a few changes in the comments as well as in the plt.xlabel and plt.ylabel.
